docs/proposals/cascading-deletion.md

```diff
@@ -0,0 +1,230 @@
+
+* Controllers: when creating child objects, controllers need to fill up the their ObjectMeta.ParentReferences field. Objects that don’t have a parent should have the namespace object as the parent.
+
+## Expected race conditions
+
+* *Controller of the parent resource creates a child object after GC has observed the deletion of the parent object*: There are two possible cases and both are well handled in this design.
+  * If parent’s orphaning policy is false, the parent object should have already been removed from the DAG, so when the child object is created, it’s added to the *Dirty Queue* and will be processed by the Garbage Processor.
+  * If parent’s orphaning policy is true, a tombstone should exist in the DAG, so the newly created child object will be added as child of the tombstone, and no further action will be taken.
+    * Note that this requires the tombstone to exist for a long enough period.
+
+* *GC deletes the child object before noticing the updates to the ParentReferences*: If this race turns out to be a problem in practice, the deletion request should be sent with ResouceVersion as a precondition.
+
+
+# Design 2. Finalizer framework + GC
+
+There are two parts in this design. 2A is about the finalizer framework; 2B is about how to use the framework to enforce the orphaning policy.
+
+## 2A. Finalizer framework
+
+## API changes
+
+```
+type ObjectMeta struct {
+	…
+	Finalizers map[string]string
+}
+```
+
+**ObjectMeta.Finalizers**: the map key is an ad hoc, explanatory name(e.g., "orphaning finalizer"), the value is an identifier for the responsible component that will remove the finalizer from the map. If the ObjectMeta.DeletionTimestamp is non-nil, entry in the map can only be removed.
+
+## New components
+
+* Finalizer:
+  * Like a controller, a finalizer is always running.
+  * A third party can develop thier own finalizer. A finalizer doesn't need to be registered with the API server.
+  * Watches for update events that meet two conditions:
+    1. the updated objects has the identifier of the finalizer in ObjectMeta.Finalizers;
+    2. ObjectMeta.DeletionTimestamp is updated from nil to non-nil.
+  * Applies the finalizing logic to the object in the update event.
+  * Removes itself from ObjectMeta.Finalizers after the finalizing logic is completed.
+  * The API server deletes the object after the last finalizer removes itself from the ObjectMeta.Finalizers field.
+  * Because it's possible for the finalizing logic to be applied multiple times, the finalizing logic has to be idempotent.
+  * If a finalizer fails to act in a timely manner, users can manually remove the finalizer from ObjectMeat.Finalizers by a regular kubectl command.
```

Comment:
@derekwaynecarr this is to answer your comment on "permanently stuck objects", what do you think?